### PR TITLE
mgr/ActivePyModules.cc: always acquire locks without holding GIL

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
```
[ RUN      ] TestImageReplayer/3.UpdateFeatures
flushed
remote_snap_id=5
local_snap_id=4, local_snap_ns=[mirror state=non-primary, complete=1, mirror_peer_uuids=, primary_mirror_uuid=ad722cff-fdc3-4be5-880e-6a79810af184, primary_snap_id=5, last_copied_object_number=1, snap_seqs={5=18446744073709551614}]
flushed
remote_snap_id=6
local_snap_id=5, local_snap_ns=[mirror state=non-primary, complete=1, mirror_peer_uuids=, primary_mirror_uuid=ad722cff-fdc3-4be5-880e-6a79810af184, primary_snap_id=6, last_copied_object_number=1, snap_seqs={6=18446744073709551614}]
flushed
remote_snap_id=7
local_snap_id=6, local_snap_ns=[mirror state=non-primary, complete=1, mirror_peer_uuids=, primary_mirror_uuid=ad722cff-fdc3-4be5-880e-6a79810af184, primary_snap_id=7, last_copied_object_number=1, snap_seqs={7=18446744073709551614}]
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/AsyncOpTracker.cc: In function 'AsyncOpTracker::~AsyncOpTracker()' thread 7f6db3ff7700 time 2020-12-22T08:54:30.003440+0000
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/AsyncOpTracker.cc: 13: FAILED ceph_assert(m_pending_ops == 0)
 ceph version Development (no_version) pacific (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x19d) [0x7f6e73673481]
 2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x165e703) [0x7f6e73673703]
 3: (AsyncOpTracker::~AsyncOpTracker()+0x45) [0x7f6e7357d6c1]
 4: (librbd::ImageWatcher<librbd::ImageCtx>::~ImageWatcher()+0x68) [0x564ca8fe4a1a]
 5: (librbd::ImageWatcher<librbd::ImageCtx>::~ImageWatcher()+0x18) [0x564ca8fe4ab2]
 6: (librbd::ImageCtx::shutdown()+0x2b) [0x564ca8fb4271]
 7: (librbd::image::CloseRequest<librbd::ImageCtx>::finish()+0x1b) [0x564ca9170e2b]
 8: (librbd::image::CloseRequest<librbd::ImageCtx>::handle_flush_image_watcher(int)+0x271) [0x564ca9170da9]
 9: (librbd::util::detail::C_CallbackAdapter<librbd::image::CloseRequest<librbd::ImageCtx>, &librbd::image::CloseRequest<librbd::ImageCtx>::handle_flush_image_watcher>::finish(int)+0x24) [0x564ca9171ce6]
 10: (Context::complete(int)+0x27) [0x564ca87b771d]
 11: (librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}::operator()() const+0x34) [0x564ca87b77ce]
 12: (void boost::asio::asio_handler_invoke<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>(librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}&, ...)+0x68) [0x564ca87e4774]
 13: (void boost_asio_handler_invoke_helpers::invoke<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}, {lambda()#1}>(librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}&, {lambda()#1}&)+0x33) [0x564ca87e4109]
 14: (void boost::asio::system_executor::dispatch<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}, std::allocator<void> >(librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}&&, std::allocator<void> const&) const+0x51) [0x564ca87e3829]
 15: (boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>::operator()()+0x52) [0x564ca87e36c2]
 16: (void boost::asio::asio_handler_invoke<boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}> >(boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&, ...)+0x68) [0x564ca87e3380]
 17: (void boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>, librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>(boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&, librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}&)+0x33) [0x564ca87e2b6c]
 18: (void boost::asio::detail::handler_work<boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>, boost::asio::system_executor, boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}> >::complete<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>(librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}&, librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}&)+0x27) [0x564ca87e127b]
 19: (boost::asio::detail::completion_handler<boost::asio::detail::work_dispatcher<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}> >::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long)+0xd6) [0x564ca87df43e]
 20: (boost::asio::detail::scheduler_operation::complete(void*, boost::system::error_code const&, unsigned long)+0x36) [0x564ca8fa652e]
 21: (boost::asio::detail::strand_service::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long)+0xa2) [0x564ca8fa6ccc]
 22: (boost::asio::detail::scheduler_operation::complete(void*, boost::system::error_code const&, unsigned long)+0x36) [0x564ca8fa652e]
 23: (boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&)+0x1c2) [0x564ca9338a26]
 24: (boost::asio::detail::scheduler::run(boost::system::error_code&)+0xf4) [0x564ca93385fa]
 25: (boost::asio::io_context::run()+0x48) [0x564ca96a11d8]
 26: (ceph::async::io_context_pool::start(short)::{lambda()#1}::operator()() const+0x1f) [0x564ca96a180f]
 27: (void std::__invoke_impl<void, ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::__invoke_other, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)+0x20) [0x564ca96a69ee]
 28: (std::__invoke_result<ceph::async::io_context_pool::start(short)::{lambda()#1}>::type std::__invoke<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::__invoke_result&&, (ceph::async::io_context_pool::start(short)::{lambda()#1}&&)...)+0x20) [0x564ca96a5ebe]
 29: (std::invoke_result<ceph::async::io_context_pool::start(short)::{lambda()#1}>::type std::invoke<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::invoke_result&&, (ceph::async::io_context_pool::start(short)::{lambda()#1}&&)...)+0x20) [0x564ca96a4f26]
 30: _ZZ17make_named_threadIZN4ceph5async15io_context_pool5startEsEUlvE_JEESt6threadSt17basic_string_viewIcSt11char_traitsIcEEOT_DpOT0_ENKUlSA_SD_E_clIS3_JEEEDaSA_SD_()
 31: (std::basic_string_view<char, std::char_traits<char> > std::__invoke_impl<void, make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}, {lambda()#1}>(std::__invoke_other, make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}&&, {lambda()#1}&&)+0x37) [0x564ca96a9020]
 32: (std::__invoke_result<std::basic_string_view<char, std::char_traits<char> >, (make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}&&)...>::type std::__invoke<make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}&&)::{lambda(auto:1, auto:2&&)#1}, {lambda()#1}>(make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}, {lambda()#1}&&)+0x37) [0x564ca96a8d8b]
 33: (void std::thread::_Invoker<std::tuple<make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}, {lambda()#1}> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>)+0x43) [0x564ca96a8b57]
 34: (std::thread::_Invoker<std::tuple<make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}, {lambda()#1}> >::operator()()+0x18) [0x564ca96a8876]
 35: (std::thread::_State_impl<std::thread::_Invoker<std::tuple<make_named_thread<ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::basic_string_view<char, std::char_traits<char> >, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)::{lambda(auto:1, auto:2&&)#1}, {lambda()#1}> > >::_M_run()+0x1c) [0x564ca96a8202]
 36: /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xd0d80) [0x7f6e71421d80]
 37: /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f6e7d4796db]
 38: clone()
```

**Review by jdurgin — APPROVED** [review]
Nice refactor! RAII is quite helpful for this.

File: src/mgr/ActivePyModules.cc [hunk 1]
```diff
@@ -56,6 +56,66 @@ ActivePyModules::ActivePyModules(PyModuleConfig &module_config_,
...
 ActivePyModules::~ActivePyModules() = default;
 
+namespace {
+  // because the Python runtime could relinquish the GIL when performing GC
+  // and reacquire it afterwards, we should enforce following locking policy:
+  // 1. do not acquire locks when holding the GIL, use a lock<PyAllowThreads>
+  //    to to guard the code which acquires non-gil locks.
+  // 2. always hold a GIL when calling python functions, for example, when
+  //    constructing a PyFormatter instance.
+  // a wrapper that provides a convenient RAII-styple mechinary for acquiring
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/mgr/ActivePyModules.cc]: @jdurgin thanks for reviewing! updated and fixed the typo in commit message as well.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-02-02_15:01:12-rados-wip-kefu-testing-2021-02-02-1939-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_08:10:00-rados-wip-kefu-testing-2021-02-03-1357-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_14:17:43-rados-wip-kefu-testing-2021-02-03-2051-distro-basic-smithi/

**Pull request merged**